### DISCUSSION TITLE
MIMXRT1170-EVK Zephyr Network Performance

### DISCUSSION THREAD

**Discussion by jacob-wienecke-nxp** [zephyrproject-rtos/zephyr#72760] [Show and tell] [answered]
# MIMXRT1170-EVK Zephyr Network Performance
## Overview
I recently had an opportunity to work with the zephyr zperf: Network Traffic Generator to achieve some performance metrics. I followed along with the documentation here: [Zperf Documentation](https://docs.zephyrproject.org/latest/connectivity/networking/api/zperf.html).

## Summary
This table summarizes the bandwidth results using the tests detailed below:

| Direction                | TCP (Mbps) | UDP (Mbps) |
| ------------------------ | ---------- | ---------- |
| Zephyr Server, PC Client | 87.7       | 95.1       |
| PC Server, Zephyr Client | 73.6       | 79.8       |

## PC Configuration
In testing, I used a windows machine with iperf-2.0.5 connected to the 100M ENET (J3) port of MIMXRT1170-EVK
Zephyr v3.6
PC configured with IP address: 192.0.2.2
Subnet Mask: 255.255.255.0

### Flashed the zperf example to run on cm7 core of RT1170:
1) 'west build -p always -b mimxrt1170_evk_cm7 samples/net/zperf'
2) 'west flash'

### Serial Terminal Input 
`zperf tcp download`

### CMD Input:
`iperf -V -c 192.0.2.1 -p 5001`

### Serial Terminal Output 
```
Duration: 10.03 s
Rate: 87.69 Mbps
```

### CMD Output
```
Interval: 0.0-10.0 sec
Transfer: 105 Mbytes
Bandwidth: 87.8 Mbits/sec
```

### Serial Terminal Input 
`zperf udp download`

### CMD Input:
`iperf -u -c 192.0.2.1 -b 100M -V`

### Serial Terminal Output 
```
Duration: 10.45 s
Received packets: 84531
nb packets lost: 0
nb packets outorder: 0
jitter 1.05 ms
rate: 95.07 Mbps
```

### CMD Output
```
Interval: 0.0 – 10.5 sec
Transfer: 119 MBytes
Bandwidth: 95.1 Mbits/sec
Jitter: 1.053 ms
Loss: 0/84531 (0%)
```

### Serial Terminal Input 
`zperf tcp upload 192.0.2.2 5001 10 1K 1M`

### CMD Input:
`iperf -s`

### Serial Terminal Output 
```
Duration: 10.00 s
Packet Size: 1000 bytes
Rate: 1000 kbps
Num packets: 92,954
Num Errors: 0
Rate: 74.36 Mbps
```

### CMD Output
```
Interval: 0.0-10.1 sec
Transfer: 88.6 Mbytes
Bandwidth: 73.6 Mbits/sec
```

### Serial Terminal Input 
`zperf udp upload 192.0.2.2 5001 10 1K 100M`

### CMD Input:
`iperf -s -u`

### Serial Terminal Output 
```
Duration: 10.00 s
Packet Size: 1000 bytes
Rate: 10 000 kbps
Num packets: 99,778
Num Packets out of order: 0
Jitter: 198 us
Rate: 79.81 Mbps
```

### CMD Output
```
Interval: 0.0-10.0 sec
Transfer: 95.2 Mbytes
Bandwidth: 79.8 Mbits/sec
Jitter 0.199 ms
Loss/Total Datagrams: 0/99778 (0%)
```

**Comment by Vijayakannan** [comment]
Hi @jacob-wienecke-nxp, Good to see the results.

I see this board (MIMXRT1170-EVKB) has,

- Ethernet (10/100/1000M) connector
- Ethernet (10/100M) connector

Have you tried similar experiment using 1000M connector with more than 100Mbps line rate?

**Reply by deribaucourt** [reply]
I've also been looking for Gbps Ethernet support, it doesn't look like the Zephyr driver currently supports it for this board.
`eth_mcux.c`:
```c
static enum ethernet_hw_caps eth_mcux_get_capabilities(const struct device *dev)
{
	ARG_UNUSED(dev);

	return ETHERNET_HW_VLAN | ETHERNET_LINK_10BASE_T |
#if defined(CONFIG_PTP_CLOCK_MCUX)
		ETHERNET_PTP |
#endif
#if defined(CONFIG_NET_DSA)
		ETHERNET_DSA_MASTER_PORT |
#endif
#if defined(CONFIG_ETH_MCUX_HW_ACCELERATION)
		ETHERNET_HW_TX_CHKSUM_OFFLOAD |
		ETHERNET_HW_RX_CHKSUM_OFFLOAD |
#endif
		ETHERNET_AUTO_NEGOTIATION_SET |
		ETHERNET_LINK_100BASE_T;
}
```

**Reply by Vijayakannan** [reply]
> I've also been looking for Gbps Ethernet support, it doesn't look like the Zephyr driver currently supports it for this board. `eth_mcux.c`:
> 
> ```c
> static enum ethernet_hw_caps eth_mcux_get_capabilities(const struct device *dev)
> {
> 	ARG_UNUSED(dev);
> 
> 	return ETHERNET_HW_VLAN | ETHERNET_LINK_10BASE_T |
> #if defined(CONFIG_PTP_CLOCK_MCUX)
> 		ETHERNET_PTP |
> #endif
> #if defined(CONFIG_NET_DSA)
> 		ETHERNET_DSA_MASTER_PORT |
> #endif
> #if defined(CONFIG_ETH_MCUX_HW_ACCELERATION)
> 		ETHERNET_HW_TX_CHKSUM_OFFLOAD |
> 		ETHERNET_HW_RX_CHKSUM_OFFLOAD |
> #endif
> 		ETHERNET_AUTO_NEGOTIATION_SET |
> 		ETHERNET_LINK_100BASE_T;
> }
> ```
Ok. I though mcux driver deprecated and the replacement is this 
drivers/ethernet/nxp_enet/eth_nxp_enet.c

![image](https://github.com/zephyrproject-rtos/zephyr/assets/11553893/5a24dca8-d891-4508-b307-257a1abf2751)

**Reply by jacob-wienecke-nxp** [reply, marked as answer]
[NXP ENET_1G gigabit ethernet on mimxrt1170_evk](https://github.com/zephyrproject-rtos/zephyr/pull/69207), though I haven't tested it recently. Early on I was seeing some performance issues with this particular PR.

**Comment by hwmaier** [comment]
For those who compare these results with other boards, the zperf example runs the code on this NXP board from 0-wait state ITCM memory. Refer to `mimxrt1170_evk_mimxrt1176_cm7_A.conf` file. I am not sure how realistic this is, as most apps would not fit into that memory space on other boards.

**Reply by dleach02** [reply]
That is true. The QSPI on our boards, while cheap, have a tradeoff in performance. We are also doing experiments to determine the subset of code that, if moved to ITCM, will give you the optimum performance.
